src/ws.nim

```diff
@@ -0,0 +1,415 @@
+import chronos, httputils, strutils, base64, std/sha1, random, http,
+        uri, times, chronos/timer, tables, stew/byteutils, eth/[keys], stew/endians2
+
+const
+  SHA1DigestSize = 20
+  WSHeaderSize = 12
+  WSOpCode = {0x00, 0x01, 0x02, 0x08, 0x09, 0x0a}
+
+type
+  ReadyState* = enum
+    Connecting = 0 # The connection is not yet open.
+    Open = 1       # The connection is open and ready to communicate.
+    Closing = 2    # The connection is in the process of closing.
+    Closed = 3     # The connection is closed or couldn't be opened.
+
+  WebSocket* = ref object
+    tcpSocket*: StreamTransport
+    version*: int
+    key*: string
+    protocol*: string
+    readyState*: ReadyState
+    masked*: bool # send masked packets
+    rng*: ref BrHmacDrbgContext
+
+  WebSocketError* = object of IOError
+
+  Base16Error* = object of CatchableError
+    ## Base16 specific exception type
+
+  HeaderFlag* {.size: sizeof(uint8).} = enum
+    rsv3
+    rsv2
+    rsv1
+    fin
+  HeaderFlags = set[HeaderFlag]
+
+  HttpCode* = enum
+    Http101 = 101 # Switching Protocols
+
+proc handshake*(ws: WebSocket, header: HttpRequestHeader) {.async.} =
+  ## Handles the websocket handshake.
+  try: ws.version = parseInt(header["Sec-WebSocket-Version"])
```

Comment:
Recently i have discovered new more appropriate procedure for int parsing. It will not raise exceptions anymore, but it still could return negative values.
https://github.com/nim-lang/Nim/blob/v1.2.6/lib/pure/parseutils.nim#L466

Reply:
Done